rccheck/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+struct DerBytesVisitor;
+
+impl<'de> Visitor<'de> for DerBytesVisitor {
+    type Value = Psk<'de>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a borrowed Subject Public Key Info in DER format")
+    }
+
+    fn visit_borrowed_bytes<E>(self, v: &'de [u8]) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        let (_, spki) = SubjectPublicKeyInfo::from_der(v).map_err(Error::custom)?;
+        Ok(Psk(spki))
+    }
+
+    fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        let (_, spki) = SubjectPublicKeyInfo::from_der(v.as_bytes()).map_err(Error::custom)?;
+        Ok(Psk(spki))
+    }
+}
+
+impl<'de> Deserialize<'de> for Psk<'de> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_bytes(DerBytesVisitor)
+    }
+}
+
+////////////////////////////////////////////////////////////////
+/// end Ser/de
+////////////////////////////////////////////////////////////////
+
+/// A `ClientCertVerifier` that will ensure that every client provides a valid, expected
+/// certificate, without any name checking.
+impl<'a> ClientCertVerifier for Psk<'a> {
+    fn offer_client_auth(&self) -> bool {
+        true
+    }
+
+    fn client_auth_mandatory(&self) -> Option<bool> {
+        Some(true)
+    }
+
+    fn client_auth_root_subjects(&self) -> Option<rustls::DistinguishedNames> {
+        // We can't guarantee subjects before having seen the cert
+        None
+    }
+
+    fn verify_client_cert(
+        &self,
+        end_entity: &rustls::Certificate,
+        intermediates: &[rustls::Certificate],
+        now: SystemTime,
+    ) -> Result<ClientCertVerified, rustls::Error> {
+        // Check this matches the key we expect
+        let cert = X509Certificate::from_der(&end_entity.0[..])
+            .map_err(|_| rustls::Error::InvalidCertificateEncoding)?;
+        let spki = cert.1.public_key().clone();
+        if spki != self.0 {
+            return Err(rustls::Error::InvalidCertificateData(format!(
+                "invalid peer certificate: received {:?} instead of expected {:?}",
+                spki, self.0
+            )));
+        }
+
+        // We now check we're receiving correctly signed data with the expected key
+        let (cert, chain, trustroots) = prepare_for_self_signed(end_entity, intermediates)?;
+        let now = webpki::Time::try_from(now).map_err(|_| rustls::Error::FailedToGetCurrentTime)?;
+        cert.verify_is_valid_tls_client_cert(
+            SUPPORTED_SIG_ALGS,
+            &webpki::TlsClientTrustAnchors(&trustroots),
+            &chain,
+            now,
+        )
+        .map_err(pki_error)
+        .map(|_| ClientCertVerified::assertion())
+    }
+}
+
+impl<'a> ServerCertVerifier for Psk<'a> {
+    fn verify_server_cert(
```

Comment:
Here I would add some docs about the high level protocol, and the verification steps withint (with numbers), then inside you can comment with reference to these numbers. Allows the crypto reviewer to first understand if the protocol is correct, and then to understand if the implementation is faithful.

Reply:
Excellent suggestion, thanks!